### PR TITLE
CSHARP-2984: Abstract getter confuses auto property mapping.

### PR DESCRIPTION
**Pull request by DmitryLukyanov**
The logic:
1. When a user configures `BsonConstructorAttribute`, it means that the marked constructor will be automatically added to the creatorMaps list even though this constructor could be skipped in the previous ImmutableTypeClassMapConvention step. 
2. This newly marked constructor can have arguments that were not added to the list of DeclaredMemberMaps in ImmutableTypeClassMapConvention step.
3. So, IPostProcessingConvention helps to check once again the arguments from constructors marked by BsonConstructorAttribute. And if they match, then we will add them to the DeclaredMemberMaps list

### CODE REVIEW

File: tests/MongoDB.Bson.Tests/Jira/CSharp1559Tests.cs [hunk 1]
```diff
@@ -14,6 +14,7 @@
 */
```
Comment (DmitryLukyanov) [inline tests/MongoDB.Bson.Tests/Jira/CSharp1559Tests.cs:14]: Maybe it makes sense to move these tests to the new file without Jira number in the name. Thoughts?

**Comment by DmitryLukyanov** [comment]
EG: https://evergreen.mongodb.com/version/5e6f69c00ae6067af1d94a8f

File: src/MongoDB.Bson/Serialization/Conventions/ImmutableTypeClassMapConvention.cs [hunk 2]
```diff
@@ -35,8 +38,7 @@ public void Apply(BsonClassMap classMap)
                 return;
             }
 
-            var propertyBindingFlags = BindingFlags.Public | BindingFlags.Instance;
-            var properties = typeInfo.GetProperties(propertyBindingFlags);
+            var properties = typeInfo.GetProperties(PropertyBindingsFlags);
```
Comment (rstam) [inline src/MongoDB.Bson/Serialization/Conventions/ImmutableTypeClassMapConvention.cs]: I think the entire bug can be fixed with just these few lines:

    var propertyBindingFlags = BindingFlags.Public | BindingFlags.Instance;
    var properties =
        typeInfo.GetProperties(propertyBindingFlags)
        .Where(p => !p.GetCustomAttributes(typeof(BsonIgnoreAttribute)).Any())
        .ToArray();

The essence of the bug was that `[BsonIgnore]` wasn't being respected by the `ImmutableTypeClassMapConvention`.

Look into it and see what you think.

Comment (rstam) [reply src/MongoDB.Bson/Serialization/Conventions/ImmutableTypeClassMapConvention.cs]: I used the following minimal repro to test the one line change above:

    namespace TestCSharp2984
    {
        public abstract class B
        {
            [BsonIgnore]
            public abstract int A { get; }
        }
    
        public class C : B
        {
            public C(int p) { P = p; }
    
            [BsonIgnore]
            public override int A => 1;
            public int P { get; }
        }
    
        public static class Program
        {
            public static void Main(string[] args)
            {
                var c = new C(1);
                var json = c.ToJson();
            }
        }
    }

**Review by rstam — CHANGES_REQUESTED** [review]

**Comment by DmitryLukyanov** [comment]
I think it's not a good idea, the issue is not only related to BsonIgnoreAttribute. See:
1. This test with BsonConstructorAttribute fails with your suggestion https://github.com/DmitryLukyanov/mongo-csharp-driver/pull/118/files#diff-6e9fa9474c0cdcaba7b731d8f52afc68R248 
2. Which is the most important, we already have well-designed areas of responsibility in the Serialization conventions. There is a special convention AttributeConvention which is responsible for attributes processing and I think only this convention should have logic to process attributes. I don't think that it's a good idea to mix this convention and ImmutableTypeClassMapConvention which should be responsible only for saving properties and constructors.
3. We already have a special mechanism IPostProcessingConvention which I consider as a place when each convention can ensure that after calling of all other conventions, the idea which was originally added via this convention still valid. Otherwise, update it or, possibly, throw an exception.

So, in two words, It sounds like my implementation fully corresponds to the original idea :)

**Comment by rstam** [comment]
I will look at it further then. I was hoping that there would be a solution that was much simpler than the rather extensive changes you have made.

**Comment by DmitryLukyanov** [comment]
I'm not sure that it can be described as `extensive` :) The only new logic here is these lines https://github.com/DmitryLukyanov/mongo-csharp-driver/pull/118/files#diff-1bcd6b847860a1520a8a7ea2b958170aR113-R117 where we just check almost the same logic as in `Apply` with the only difference that now we consider only properties which were not added before and which present in the creators.

**Comment by rstam** [comment]
I know extensive is relative.

I was hoping that a one line change would fix it. Can't get any easier than that.

**Review by rstam — CHANGES_REQUESTED** [review]
I tried my minimal repro with your changes and got an incorrect result:

    json == "{ }"

So more work is required.

**Comment by DmitryLukyanov** [comment]
Thanks, I will check it

File: src/MongoDB.Bson/Serialization/Conventions/ImmutableTypeClassMapConvention.cs [hunk 3]
```diff
@@ -105,5 +135,30 @@ private bool CanWrite(PropertyInfo propertyInfo)
...
             return propertyInfo.CanWrite && (propertyInfo.SetMethod?.IsPublic ?? false);
         }
+
+        private PropertyInfo[] GetProperties(TypeInfo typeInfo)
+        {
+            var propertyBindingsFlags = BindingFlags.Public | BindingFlags.Instance;
+
+            return
+                typeInfo.GetProperties(propertyBindingsFlags)
+                .Where(p => p.GetCustomAttribute<BsonIgnoreAttribute>() == null)
```
Comment (DmitryLukyanov) [inline src/MongoDB.Bson/Serialization/Conventions/ImmutableTypeClassMapConvention.cs]: I don't like this change since we split attributes processing from one convention into at least two. But it's the easiest way, each other good looking way which I tried required much more changes. But I would consider more good looking refactoring for this case in the scope of a new ticket.
Also, it looks like a similar change should be considered for other conventions, at least for `ReadWriteMemberFinderConvention`

**Review by DmitryLukyanov — COMMENTED** [review]
I've merged my version and Robert's suggestion. But I still think that it contradicts to the original idea and it's better to schedule more good looking refactroing for this case. Thoughts?

File: src/MongoDB.Bson/Serialization/Conventions/ImmutableTypeClassMapConvention.cs (hunk 3 above)
Comment (rstam) [reply src/MongoDB.Bson/Serialization/Conventions/ImmutableTypeClassMapConvention.cs]: So the general pattern of the conventions working together is:

1. IClassMapConvention(s) add or remove members from the class map
2. IMemberMapConventions(s) configure member maps added in step 1
3. ICreateMapConvention(s) configure any creator maps added in step 1
4. IPostProcessingConvention(s) do any clean up on the class map

Why don't you try limiting the processing this convention does in the IClassMapConvention to only adding members to the class map, and postpone adding the constructor to the class map until the IPostProcessingConvention. That way the AttributeConventionPack will get a change to remove the `[BsonIgnore]` elements this class added because it doesn't (or shouldn't) know about the `[BsonIgnore]` attribute.

We don't usually add creators in the post processing step, but that might be what's needed here?

**Review by rstam — CHANGES_REQUESTED** [review]
See last comment in conversation tab for requested change.

File: src/MongoDB.Bson/Serialization/Conventions/ImmutableTypeClassMapConvention.cs (hunk 3 above)
Comment (rstam) [reply src/MongoDB.Bson/Serialization/Conventions/ImmutableTypeClassMapConvention.cs]: > Also, it looks like a similar change should be considered for other conventions, at least for ReadWriteMemberFinderConvention

I don't think so. The attribute conventions run after the `ReadWriteMemberFinderConvention` and will remove the members tagged with `[BsonIgnore]`.

Comment (DmitryLukyanov) [reply src/MongoDB.Bson/Serialization/Conventions/ImmutableTypeClassMapConvention.cs]: >Why don't you try limiting the processing this convention does in the IClassMapConvention to only adding members to the class map, and postpone adding the constructor to the class map until the IPostProcessingConvention. That way the AttributeConventionPack will get a change to remove the [BsonIgnore] elements this class added because it doesn't (or shouldn't) know about the [BsonIgnore] attribute.

because of `NamedParameterCreatorMapConvention` which is called here: https://github.com/mongodb/mongo-csharp-driver/blob/master/src/MongoDB.Bson/Serialization/Conventions/ConventionRunner.cs#L69. This convention configures missed arguments for ctor and it requires having already configured ctorMap

Comment (DmitryLukyanov) [reply src/MongoDB.Bson/Serialization/Conventions/ImmutableTypeClassMapConvention.cs]: >I don't think so. The attribute conventions run after the ReadWriteMemberFinderConvention and will remove the members tagged with [BsonIgnore]

agreed

**Comment by DmitryLukyanov** [comment]
Latest update from slack:
the main idea of this ticket that we don't always correctly process BsonConstructor and BsonIgnore attributes in immutable classes. The current conventions processing is not so smart to consider all cases. Previously I added a solution for this, but it included processing BsonIgnore attribute in ImmutableConvention, which is not expected since we have a specialized AttributeConvention . Also, it required adding logic into PostProcessing step and we're not sure whether it's a good idea or no.
I see several better solutions to fix it, but all of them require changes in the base logic for convention processing and I'm not sure that it's super safe to change it.
So I see 3 options:
Close ticket with the comment that we don't support some properties configurations for immutable classes.
Leave the currently added solution. I think processing BsonIgnore in ImmutableConvention is not super good-looking but at the same time, I think it's harmless.
Adding more significant changes into convention processing. I think there is a way to make them safe (at least it should be)

**Review by rstam — CHANGES_REQUESTED** [review]
While trying to understand the changes made I single stepped through the debugger and had an insight into a possible much simpler fix.

See:

https://github.com/rstam/mongo-csharp-driver/tree/csharp2984-1005

The change consists of reverting ImmutableTypeClassMapConvention back to its original form and then commenting out these 4 lines:

https://github.com/rstam/mongo-csharp-driver/commit/c242d15f476218d6e524c61de69e16c8e870dfab#diff-1bcd6b847860a1520a8a7ea2b958170aR55

If you agree that this simple fix is sufficient then these lines should actually be deleted instead of commented out.

The rationale behind deleting the lines is this: IF they declared a constructor that takes fewer arguments than would seem to be necessary then that is because they are saying it is OK to use this constructor. If the document being deserialized has all the fields necessary to call a different constructor with more arguments then we would choose the constructor with more arguments during deserialization. We would only choose the constructor with fewer arguments if the document being deserialized did not have enough fields to call a different constructor with more arguments.

p.s. All your new tests still pass even with the simpler suggested fix.

**Comment by rstam** [comment]
Note: I'd also refactor the new tests in the CSharp2984Tests file to assert that the result of calling `ToJson` is what is expected.

Currently the tests assert that after a round trip (ToJson => Deserialize) the resulting POCO is correct, but it would probably be a good idea to verify that the JSON is correct also.

**Comment by DmitryLukyanov** [comment]
I would say that the suggested fix is not really safe. Now, we will consider more classes as `immutable` and as a result, these new cases will be processed in a more difficult code path. Then, I think we will have more Jira tickets with regressions ideas from users. I would suggest to avoid this.

**Comment by rstam** [comment]
I don't think the suggested fix is not safe. Can you provide an example class that would be mapped incorrectly?

**Review by rstam — CHANGES_REQUESTED** [review]
Changing status to changes requested.

**Comment by DmitryLukyanov** [comment]
>I don't think the suggested fix is not safe. 

For example this one:

        public class Base
        {
            public int C { get; }
        }

        public class Test : Base
        {
            public Test(int a)
            {
                A = a;
            }

            [BsonIgnore]
            public int A { get; }
        }

it's definitely not the most realistic class, but it's the only one which I found. I'm almost sure there are more cases than this.
Not sure how better we can solve this, but it would be great to avoid regressions questions as we had before :)

**Comment by rstam** [comment]
The Test class you defined is impossible to instantiate (during deserialization).

You used `[BsonIgnore]` to suppress serialization of the `A` property.

That means that during deserialization there is no value to pass to the constructor.

I don't think there is any reason to expect our serialization machinery to work with this class.

**Review by rstam — CHANGES_REQUESTED** [review]
See comment on Conversation tab.

**Comment by DmitryLukyanov** [comment]
discussed offline. I think Robert's suggestion is not 100% safe and may lead to regression issues. So, let's consider this change in the scope of CSHARP-3230 where we can make a more complex solution.

**Comment by rstam** [comment]
The gist of the bug is that the ImmutableTypeClassMapConvention is skipping a constructor it shouldn't here:

https://github.com/mongodb/mongo-csharp-driver/blob/master/src/MongoDB.Bson/Serialization/Conventions/ImmutableTypeClassMapConvention.cs#L57

My initial proposal is to delete lines 55-58. The argument is that there is no reason to check the number of parameters. If the constructor exists it is by definition safe to use.

But Dmitry is not comfortable with that for some reason.

For a more limited fix that addresses this sample class only see:

https://github.com/rstam/mongo-csharp-driver/tree/csharp2984-1023

The idea of this commit is to keep the unnecessary test that checks the number of parameters, but to allow the presence of a [BsonConstructor] attribute to override the test.

I think this more limited fix is insufficient. But it would at least allow this ticket to be closed. On the other hand it might be a bad idea to close this ticket if we don't properly fix the issue.

**Comment by DmitryLukyanov** [comment]
>The gist of the bug is that the ImmutableTypeClassMapConvention is skipping a constructor it shouldn't here:

I don't think that this is a fully correct description. The gist of the problem that we process attributes configuration (all of them and not only BsonConstructor) in the last step in the conventions conveyor. In other words, all our possible assumptions (including further classMap validation) about class structure can be broken based on the added attributes. So, my fix suggestion was exactly about fixing possible misunderstandings between our code assumptions and the real creators and field configuration after the time when attributes were handled. That's why I used `IPostProcessingConvention`, because these compensation steps should be called after the main conventions conveyer. I agree that it's not the best solution and the "best" fix should be about avoiding the whole problem related to the fact that attribute processing is the last step which is out of the control from the previous conventions. 
I think we should handle attributes as the first steps and modify `classMap` type in a way when we don't use System.Type's properties directly (because we cannot modify type state at all). 
Instead (but it's not easy and should be additionally investigated in detail), we should use our abstractions which will be created based on the class type in the first step and be modified based on the attributes in the second step. All next conventions will use whatever our new `System.Type`-like class without considering any attributes or real class properties. That's why I mentioned the scope of https://jira.mongodb.org/browse/CSHARP-3230 for this target.

As for Robert's suggestion: 
`csharp2984-1023` - doesn't cover all cases and looks like a band-aid just to close 2984 rather than fixing the entire problem
`csharp2984-1005 ` - it's behavior change which looks dangerous to me. However, I can approve it and rely on Robert's opinion if he's sure that it's safe. But even if we consider it as save, it looks like not directly related to fixing the underlying problem here.

cc: @rstam , @jyemin

**Comment by rstam** [comment]
> As for Robert's suggestion:
csharp2984-1023 - doesn't cover all cases and looks like a band-aid just to close 2984 rather than fixing the entire problem
csharp2984-1005 - it's behavior change which looks dangerous to me. However, I can approve it and rely on Robert's opinion if he's sure that it's safe. But even if we consider it as save, it looks like not directly related to fixing the underlying problem here

Agreed, let's not even consider csharp2984-1023. It was a thought experiment for solving 2984 ONLY with no further considerations.

As far as your statement that csharp2984-1005 does not directly fix the underlying problem... I think we have to ask WHAT underlying problem(s) we are trying to solve.

csharp2984-1005 solves ONE underlying problem, namely that the ImmutableTypeClassMapConvention incorrectly skips some constructors that should not be skipped. This is THE underlying problem with the immutable class provided as an example for CSHARP-1984.

You identify OTHER underlying problems (such as conflicts between conventions and attributes) that are more like design problems and should be addressed in the future. Changing the overall design of conventions and attributes sounds like a task for a major release.

**Comment by DmitryLukyanov** [comment]
>As far as your statement that csharp2984-1005 does not directly fix the underlying problem... I think we have to ask WHAT underlying problem(s) we are trying to solve.

misunderstanding/inconsistency between attributes convention and convention for immutable classes. They actually may conflict with each other.

>csharp2984-1005 solves ONE underlying problem, namely that the ImmutableTypeClassMapConvention incorrectly skips some constructors that should not be skipped. This is THE underlying problem with the immutable class provided as an example for CSHARP-1984.

this is one case of an underlying problem, I believe we can find more of them. Since from the one side, we don't control how/where/which attributes can be set, and from the other side, we have some rules which fields/constructors can be added via an immutable convention (not only?). So the first side can break the second one.

> like design problems and should be addressed in the future

true

**Comment by rstam** [comment]
@DmitryLukyanov can you try and find a class where my csharp2984-1005 branch suggestion could be considered a breaking change?

I don't think it is a breaking change, but I will try to find an example also, but you might be more likely to find an example (if it exists) since you are the one with the concern.

**Comment by DmitryLukyanov** [comment]
EG: https://evergreen.mongodb.com/version/5fa00f3c3e8e867645ad5cf9
I applied Robert's suggestion, ready for review

File: tests/MongoDB.Bson.Tests/Serialization/SerializationTests.cs [hunk 4]
```diff
@@ -14,30 +14,44 @@
...
 using MongoDB.Bson.Serialization;
 using MongoDB.Bson.Serialization.Attributes;
 using MongoDB.Bson.TestHelpers;
 using Xunit;
 
-namespace MongoDB.Bson.Tests.Jira
+namespace MongoDB.Bson.Tests.Serialization
 {
-    public class CSharp1559Tests
+    public class SerializationTests
```
Comment (rstam) [inline tests/MongoDB.Bson.Tests/Serialization/SerializationTests.cs]: Why is this file renamed? And if we want to rename it we need a better name ("SerializationTests" is too general).

File: tests/MongoDB.Bson.Tests/Serialization/SerializationTests.cs [hunk 5]
```diff
@@ -234,5 +248,126 @@ public DerivedWithoutSetter_IntermediateBaseAndWithProtectedConstructor(int? x,
...
+                Y = y;
+            }
+
+            [BsonIgnore]
+            public override int X { get; } = 1;
+
+            public int Y { get; } = 0; // should be overwritten
+        }
+
+        public class ClassWithTwoConstructorsWhereTheFirstIfNotFullyMatchedAndTheSecondIsFullyMatched
```
Comment (rstam) [inline tests/MongoDB.Bson.Tests/Serialization/SerializationTests.cs]: I think you meant `...IsNotFullyMatched...` instead of `...IfNotFullyMatched...`

**Review by rstam — CHANGES_REQUESTED** [review]
Question about file that was renamed.

Will also continue reviewing the other changes.

**Review by JamesKovacs — APPROVED** [review]
LGTM

File: tests/MongoDB.Bson.Tests/Serialization/SerializationTests.cs (hunk 4 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Bson.Tests/Serialization/SerializationTests.cs]: I will revert it

File: tests/MongoDB.Bson.Tests/Serialization/SerializationTests.cs [hunk 6]
```diff
@@ -14,30 +14,44 @@
...
+        [InlineData(typeof(DerivedImmutableWithMorePropertiesThanInConstructorButWithBsonElementAttributeAndWithBsonConstructorAttribute_AbstractBaseImmutable), new[] { 2 }, "{ \"X\" : 1, \"Y\" : 2 }")]
+        [InlineData(typeof(DerivedImmutableWithMorePropertiesThanInConstructorButWithBsonElementAttributeAndWithBsonConstructorAttribute_AbstractBaseImmutableWithConstructor), new[] { 1, 2 }, "{ \"X\" : 1, \"Y\" : 2 }")]
+        [InlineData(typeof(DerivedImmutableWithMorePropertiesThanInConstructorButWithBsonIgnoreAttribute_AbstractBaseImmutableWithAbstractProperty), new[] { 2 }, "{ \"Y\" : 2 }")]
+        [InlineData(typeof(DerivedImmutableWithMorePropertiesThanInConstructorButWithBsonIgnoreAttribute_AbstractBaseImmutableWithConstructor), new[] { 1 }, "{ \"X\" : 1 }")]
+        [InlineData(typeof(ClassWithTwoConstructorsWhereTheFirstIfNotFullyMatchedAndTheSecondIsFullyMatched), new[] { 1, 2 }, "{ \"X\" : 1, \"Y\" : 2 }")]
+        [InlineData(typeof(ClassWithOneConstructorThatIsNotFullyMatched), new[] { 1, 2 }, "{ \"X\" : 1, \"Y\" : 2, \"Z\" : null }")]
```
Comment (rstam) [inline tests/MongoDB.Bson.Tests/Serialization/SerializationTests.cs]: It's worth asking whether `Z` should be serialized or not.

We might want to make a small change to `ImmutableTypeClassMapConvention` (line 88) so that only properties that can be matched to a constructor argument (any constructor) will be mapped.

This used to not be an issue because we used to only consider the one constructor (if any) that had the same number of arguments as the class had properties.

**Review by rstam — CHANGES_REQUESTED** [review]

File: tests/MongoDB.Bson.Tests/Jira/CSharp1559Tests.cs [hunk 7]
```diff
@@ -26,18 +27,31 @@ namespace MongoDB.Bson.Tests.Jira
...
+        [InlineData(typeof(DerivedImmutableWithMorePropertiesThanInConstructorButWithBsonElementAttributeAndWithBsonConstructorAttribute_AbstractBaseImmutableWithConstructor), new[] { 1, 2 }, "{ \"X\" : 1, \"Y\" : 2 }")]
+        [InlineData(typeof(DerivedImmutableWithMorePropertiesThanInConstructorButWithBsonIgnoreAttribute_AbstractBaseImmutableWithAbstractProperty), new[] { 2 }, "{ \"Y\" : 2 }")]
+        [InlineData(typeof(DerivedImmutableWithMorePropertiesThanInConstructorButWithBsonIgnoreAttribute_AbstractBaseImmutableWithConstructor), new[] { 1 }, "{ \"X\" : 1 }")]
+        [InlineData(typeof(ClassWithTwoConstructorsWhereTheFirstIsNotFullyMatchedAndTheSecondIsFullyMatched), new[] { 1, 2 }, "{ \"X\" : 1, \"Y\" : 2 }")]
+        [InlineData(typeof(ClassWithOneConstructorThatIsNotFullyMatched), new[] { 1, 2 }, "{ \"X\" : 1, \"Y\" : 2, \"Z\" : null }")]
+        public void Serialization_should_return_expected_result(Type testCaseType, int[] arguments, string expectedJson)
         {
-            var testCase = Activator.CreateInstance(testCaseType, 1, 2);
+            var testCase = Activator.CreateInstance(testCaseType, arguments.Select(a => (object)a).ToArray());
```
Comment (rstam) [inline tests/MongoDB.Bson.Tests/Jira/CSharp1559Tests.cs:48]: Not neccesary to change but this can be written slightly simpler as:

    var testCase = Activator.CreateInstance(testCaseType, arguments.Cast<object>().ToArray())

File: tests/MongoDB.Bson.Tests/Jira/CSharp1559Tests.cs [hunk 8]
```diff
@@ -26,18 +27,31 @@ namespace MongoDB.Bson.Tests.Jira
...
+        [InlineData(typeof(DerivedImmutableWithMorePropertiesThanInConstructorButWithBsonElementAttributeAndWithBsonConstructorAttribute_AbstractBaseImmutable), new[] { 2 }, "{ \"X\" : 1, \"Y\" : 2 }")]
+        [InlineData(typeof(DerivedImmutableWithMorePropertiesThanInConstructorButWithBsonElementAttributeAndWithBsonConstructorAttribute_AbstractBaseImmutableWithConstructor), new[] { 1, 2 }, "{ \"X\" : 1, \"Y\" : 2 }")]
+        [InlineData(typeof(DerivedImmutableWithMorePropertiesThanInConstructorButWithBsonIgnoreAttribute_AbstractBaseImmutableWithAbstractProperty), new[] { 2 }, "{ \"Y\" : 2 }")]
+        [InlineData(typeof(DerivedImmutableWithMorePropertiesThanInConstructorButWithBsonIgnoreAttribute_AbstractBaseImmutableWithConstructor), new[] { 1 }, "{ \"X\" : 1 }")]
+        [InlineData(typeof(ClassWithTwoConstructorsWhereTheFirstIsNotFullyMatchedAndTheSecondIsFullyMatched), new[] { 1, 2 }, "{ \"X\" : 1, \"Y\" : 2 }")]
+        [InlineData(typeof(ClassWithOneConstructorThatIsNotFullyMatched), new[] { 1, 2 }, "{ \"X\" : 1, \"Y\" : 2, \"Z\" : null }")]
```
Comment (rstam) [inline tests/MongoDB.Bson.Tests/Jira/CSharp1559Tests.cs]: This comment was lost when you renamed the file:

It's worth asking whether Z should be serialized or not.

We might want to make a small change to ImmutableTypeClassMapConvention (line 88) so that only properties that can be matched to a constructor argument (any constructor) will be mapped.

This used to not be an issue because we used to only consider the one constructor (if any) that had the same number of arguments as the class had properties.

**Review by rstam — CHANGES_REQUESTED** [review]
Whether changes are needed or not depend on discussing my comment about whether `Z` should be part of the serialized form for the test case on line 44.

File: tests/MongoDB.Bson.Tests/Jira/CSharp1559Tests.cs (hunk 8 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Bson.Tests/Jira/CSharp1559Tests.cs]: >It's worth asking whether Z should be serialized or not.

I think it should be serialized. Because the class can be written like this:

        public class ClassWithOneConstructorThatIsNotFullyMatched
        {
            public int? X { get; }
            public int? Y { get; }
            public int? Z { get; } = 2;  // The default value

            public ClassWithOneConstructorThatIsNotFullyMatched(int? x, int? y)
            {
                X = x;
                Y = y;
            }
        }

and in this case, we will have this serialized value: `{X=1, Y=2, Z=2}.`

File: tests/MongoDB.Bson.Tests/Jira/CSharp1559Tests.cs (hunk 7 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Bson.Tests/Jira/CSharp1559Tests.cs:48]: Yep, I will consider this way for next time :)

File: tests/MongoDB.Bson.Tests/Jira/CSharp1559Tests.cs (hunk 8 above)
Comment (rstam) [reply tests/MongoDB.Bson.Tests/Jira/CSharp1559Tests.cs]: The counter argument is that it probably won't round trip because `Z` in the document can't be matched to any constructor argument.

**Review by rstam — CHANGES_REQUESTED** [review]

File: tests/MongoDB.Bson.Tests/Jira/CSharp1559Tests.cs (hunk 8 above)
Comment (rstam) [reply tests/MongoDB.Bson.Tests/Jira/CSharp1559Tests.cs]: When reading the document the Z value will be ignored (even though I expected an exception to be thrown).

So the following executes without an exception:

    var json = "{ X : 1, Y : 2, Z : 123456 }"
    var result = BsonSerializer.Deserialize(json, testCaseType);

BUT... the Z value of the result will be `2` (or whatever the class sets it to).

I don't think we want to write values OUT to the database that we can't read back IN.

Comment (rstam) [reply tests/MongoDB.Bson.Tests/Jira/CSharp1559Tests.cs]: I'm willing to ignore this issue for now if you and @JamesKovacs both argue in favor of ignoring it.

**Review by rstam — CHANGES_REQUESTED** [review]

File: tests/MongoDB.Bson.Tests/Jira/CSharp1559Tests.cs (hunk 8 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Bson.Tests/Jira/CSharp1559Tests.cs]: Maybe we need to allow saving all fields and not only mentioned in the constructor then? But it will require additional investigation, I can create a ticket to do it

Comment (rstam) [reply tests/MongoDB.Bson.Tests/Jira/CSharp1559Tests.cs]: One of our goals in serialization is to no silently lose data.

To achieve that goal we should NOT include `Z` in the serialized form because it is ONE WAY.

I don't think a new ticket is needed... but would like to hear @JamesKovacs  opinion as well.

Comment (rstam) [reply tests/MongoDB.Bson.Tests/Jira/CSharp1559Tests.cs]: Note that we already have precedent (in mutable classes) for NOT serializing properties that can't be read back in.

**Review by rstam — CHANGES_REQUESTED** [review]

**Comment by rstam** [comment]
This is the minor change I am proposing:

https://github.com/rstam/mongo-csharp-driver/tree/csharp2984-1102

**Review by JamesKovacs — CHANGES_REQUESTED** [review]
We should handle this edge case upfront with this PR. If we leave it for later, we risk serializing properties such as Z that will fail deserialization once we fix the edge case in a future release. @rstam's suggestion of ignoring properties that cannot be read in again seems like a reasonable way forward. I think the naming of the helper method can be improved for clarity, but the technique is sound.

**Comment by rstam** [comment]
Here's some background motivation for where I'm coming from.

Currently this mutable class:

    public class MutablePerson
    {
        public string FirstName { get; set; }
        public string LastName { get; set; }
        public string FullName => FirstName + " " + LastName;
    }

serializes as:

    "{ \"FirstName\" : \"John\", \"LastName\" : \"Doe\" }"

and it seems reasonable that the equivalent immutable class:

    public class ImmutablePerson
    {
        public ImmutablePerson(string firstName, string lastName)
        {
            FirstName = firstName;
            LastName = lastName;
        }

        public string FirstName { get; }
        public string LastName { get; }
        public string FullName => FirstName + " " + LastName;
    }

should serialize the SAME way.

**Comment by rstam** [comment]
> I think the naming of the helper method can be improved for clarity, but the technique is sound.

How about `PropertyHasMatchingConstructorArgument`

**Comment by JamesKovacs** [comment]
`PropertyHasMatchingConstructorArgument` is much more intent-revealing. I like it.

**Comment by rstam** [comment]
I made some relatively minor fixes to https://github.com/rstam/mongo-csharp-driver/tree/csharp2984-1102 and now all tests are passing.

My suggested changes to ImmutableTypeClassMapConvention can be summarized succinctly as:

1. Consider all public constructors (not just the one with the most parameters)
2. Only map properties that match some constructor argument

These changes to ImmutableTypeClassMapConvention are relatively trivial. Only the second has some complexity in detecting whether there is a match or not.

File: src/MongoDB.Bson/Serialization/Conventions/ImmutableTypeClassMapConvention.cs [hunk 9]
```diff
@@ -43,19 +45,14 @@ public void Apply(BsonClassMap classMap)
             }
 
             var anyConstructorsWereFound = false;
-            var constructorBindingFlags = BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.Instance;
-            foreach (var ctor in typeInfo.GetConstructors(constructorBindingFlags))
+            foreach (var ctor in typeInfo.GetConstructors(_constructorBindingFlags))
```
Comment (rstam) [inline src/MongoDB.Bson/Serialization/Conventions/ImmutableTypeClassMapConvention.cs]: I find moving the binding flag constants to a non-local variable (field in this case) unhelpful.

It means we can't fully understand what line 48 does without tracking down what the *current* value of the variable is.

I would never write code this way but I am willing to LGTM this way if you insist.

File: src/MongoDB.Bson/Serialization/Conventions/ImmutableTypeClassMapConvention.cs [hunk 10]
```diff
@@ -105,5 +106,48 @@ private bool CanWrite(PropertyInfo propertyInfo)
...
+                {
+                    if (PropertyMatchesSomeConstructorParameter(constructorInfo))
+                    {
+                        return true;
+                    }
+                }
+            }
+
+            // also map properties that match some constructor parameter that might be called by a derived class
+            foreach (var constructorInfo in classMap.ClassType.GetTypeInfo().GetConstructors(_constructorBindingFlags))
```
Comment (rstam) [inline src/MongoDB.Bson/Serialization/Conventions/ImmutableTypeClassMapConvention.cs]: I find moving the binding flag constants to a variable unhelpful. 

It means we can't fully understand what line 124 does without tracking down what the *current* value of the variable is.

File: src/MongoDB.Bson/Serialization/Conventions/ImmutableTypeClassMapConvention.cs [hunk 11]
```diff
@@ -105,5 +106,48 @@ private bool CanWrite(PropertyInfo propertyInfo)
...
+            {
+                if (classMap.ClassType.GetTypeInfo().IsAbstract || 
+                    constructorInfo.IsFamily || // protected
+                    constructorInfo.IsFamilyOrAssembly) // protected internal
```
Comment (rstam) [inline src/MongoDB.Bson/Serialization/Conventions/ImmutableTypeClassMapConvention.cs:134]: Nice catch of another uncommon (but valid) edge case.

**Review by rstam — CHANGES_REQUESTED** [review]
You didn't take all the changes I proposed in my 3240 pull request.

For one example, you chose not to test that CSHARP-2984 is fixed.

It might have been better to switch to the 3240 pull request to avoid transcription errors like this when copying code from one pull request to the other, but since you insist on using this 8-month old pull request here we are.

File: src/MongoDB.Bson/Serialization/Conventions/ImmutableTypeClassMapConvention.cs (hunk 9 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Bson/Serialization/Conventions/ImmutableTypeClassMapConvention.cs]: I would say it's helpful, it means that all searching for constructors in this file work in the same way regardless of what lines in this file we're looking in

File: src/MongoDB.Bson/Serialization/Conventions/ImmutableTypeClassMapConvention.cs (hunk 10 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Bson/Serialization/Conventions/ImmutableTypeClassMapConvention.cs]: I would say it's helpful, it means that all searching for constructors in this file work in the same way regardless of what lines in this file we're looking in

File: src/MongoDB.Bson/Serialization/Conventions/ImmutableTypeClassMapConvention.cs (hunk 9 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Bson/Serialization/Conventions/ImmutableTypeClassMapConvention.cs]: if you think that's not informative, we can change the current variable name, but I would say it's not necessary

Comment (rstam) [reply src/MongoDB.Bson/Serialization/Conventions/ImmutableTypeClassMapConvention.cs]: I agree that it is helpful in that one sense, that we have one place where we define the binding flags to be used when we call GetConstructors in two different places.

The part that is not helpful is that now you can't understand line 48 without chasing down the definition of the variable. And the fact that it is a variable means I have to figure out what the current value is (in case it might have changed).

If you insist on this at least make it a constant:

    private const BindingFlags __constructorBindingFlags = BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.Instance;

If `const` doesn't work make it `static readonly`.

Perhaps a better alternative to a global constant would be a helper method to get the constructors and then call the same helper method from both places.

File: src/MongoDB.Bson/Serialization/Conventions/ImmutableTypeClassMapConvention.cs (hunk 10 above)
Comment (rstam) [reply src/MongoDB.Bson/Serialization/Conventions/ImmutableTypeClassMapConvention.cs]: See comment above

**Review by rstam — CHANGES_REQUESTED** [review]
Are you going to pull in the tests for 2984 from my 3240 PR?

**Comment by DmitryLukyanov** [comment]
https://evergreen.mongodb.com/version/5faaeba132f4174fdd883003

File: src/MongoDB.Bson/Serialization/Conventions/ImmutableTypeClassMapConvention.cs (hunk 9 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Bson/Serialization/Conventions/ImmutableTypeClassMapConvention.cs]: done

**Review by JamesKovacs — APPROVED** [review]
LGTM

**Review by rstam — APPROVED** [review]
LGTM
